torch/testing/_internal/common_utils.py

```diff
@@ -408,33 +420,29 @@ def wrapper(*args, **kwargs):
 def wrapDeterministicFlagAPITest(fn):
     @wraps(fn)
     def wrapper(*args, **kwargs):
-        deterministic_restore = torch.is_deterministic()
+        deterministic_guard = DeterministicGuard()
 
-        is_cuda10_2_or_higher = (
-            (torch.version.cuda is not None)
-            and ([int(x) for x in torch.version.cuda.split(".")] >= [10, 2]))
-
-        if is_cuda10_2_or_higher:
+        class CuBLASConfigGuard:
             cublas_var_name = 'CUBLAS_WORKSPACE_CONFIG'
-            cublas_config_restore = os.environ.get(cublas_var_name)
-            os.environ[cublas_var_name] = ':4096:8'
-
-        def restore():
-            torch.set_deterministic(deterministic_restore)
-            if is_cuda10_2_or_higher:
-                cur_cublas_config = os.environ.get(cublas_var_name)
-                if cublas_config_restore is None:
-                    if cur_cublas_config is not None:
-                        del os.environ[cublas_var_name]
-                else:
-                    os.environ[cublas_var_name] = cublas_config_restore
-        try:
-            fn(*args, **kwargs)
-        except RuntimeError:
-            restore()
-            raise
-        else:
-            restore()
+
+            def __init__(self):
```

Comment:
Again I don't think we want to model a decorator as using dunder init and dunder del. There are some examples of different kinds of Pythonic decorators in https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_device_type.py that may be interesting to look at.

Reply:
I've changed `CuBLASConfigGuard` to act like a context manager